### DISCUSSION TITLE
iKinGazeCtrl: difference between setStabilizationMode and setTrackingMode

### DISCUSSION THREAD

**Discussion by valeriosperati** [robotology/community#226] [Q&A] [answered]
Hi,

I'm running the tutorial of `iKinGazeCtrl` on the iCubSim (`tutorial_gaze_interface`).

Here it is used the function `setTrackingMode`, which should compensate gaze when the torso is moved.
In the [webpage description](http://wiki.icub.org/brain/icub_gaze_interface.html) it is described the function `setStabilizationMode`, which as I understand, should do the same work.

It is not clear to me:
1. the difference between them **if the robot does not walk**;
2. if one function excludes the other.

Thanks, Valerio

**Comment by pattacini** [comment, marked as answer]
Hi @valeriosperati 

The two modalities elicit two different pathways as described below.

#### Tracking
When the robot has attained its target, the tracking mode simply **triggers the controller again** to keep the focus on the desired location as soon as the distance between the current fixation point and the target overcomes a given threshold.

Two things to bear in mind: 
- The target position flows through the entire controller's chain, including the Cartesian solver.
- The retrigger may happen because the target has moved or e.g. the torso is being used for other tasks.

#### Stabilization
To stabilize the gaze we can use the **gyro readouts**, which ─ differently from the case above ─ do not feed the whole controller's chain but rather act more downstream (on the **velocities**) as shown in the sketch below (details can be found in [1] and [2]).

![capture](https://cloud.githubusercontent.com/assets/3738070/26669213/4b133fb2-46ad-11e7-9884-abf564096fd7.PNG)

Therefore, the **response due to the gyro is much faster** than in the case of the compensation merely based on the position error (i.e, `target-current_fixation_point`) seen by the controller.

Of course, **you can use both modalities at the same time**.

To appreciate the benefit you can gain by enabling the stabilization, watch this video:

[![download](https://cloud.githubusercontent.com/assets/3738070/26671321/4bfc1d74-46b5-11e7-8889-3040091cec8b.png)](https://www.youtube.com/watch?v=I4ZKfAvs1y0&t=1m33s)

When stabilized by means of gyro feedback, the head does not lag behind the target.

---
#### References
[1] "[_Gaze stabilization for humanoid robots: A comprehensive framework_](https://doi.org/10.1109/HUMANOIDS.2014.7041369)", A. Roncone, U. Pattacini, G. Metta, L. Natale, **IEEE-RAS International Conference on Humanoid Robots**, Madrid, Spain, 2014.

[2] "[_A Cartesian 6-DoF Gaze Controller for Humanoid Robots_](https://doi.org/10.15607/RSS.2016.XII.022)", A. Roncone, U. Pattacini, G. Metta, L. Natale, **Proceedings of Robotics: Science and Systems**, Ann Arbor MI, 2016.

**Comment by valeriosperati** [comment]
Thanks a lot!
